In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import OneHotEncoder,StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

In [2]:
ibm = pd.read_csv("C:/Users/Arhum/Downloads/WA_Fn-UseC_-HR-Employee-Attrition.csv")
ibm

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [3]:
ibm.isnull().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

In [4]:
ibm["Attrition"] = ibm["Attrition"].map({"Yes": 1, "No": 0})

In [5]:
ibm["Attrition"].value_counts()

Attrition
0    1233
1     237
Name: count, dtype: int64

In [6]:
At_yes=ibm[ibm.Attrition==1]
At_no=ibm[ibm.Attrition==0]
pd.DataFrame(At_yes)

Age  Attrition     BusinessTravel  DailyRate              Department  \
0      41          1      Travel_Rarely       1102                   Sales   
2      37          1      Travel_Rarely       1373  Research & Development   
14     28          1      Travel_Rarely        103  Research & Development   
21     36          1      Travel_Rarely       1218                   Sales   
24     34          1      Travel_Rarely        699  Research & Development   
...   ...        ...                ...        ...                     ...   
1438   23          1  Travel_Frequently        638                   Sales   
1442   29          1      Travel_Rarely       1092  Research & Development   
1444   56          1      Travel_Rarely        310  Research & Development   
1452   50          1  Travel_Frequently        878                   Sales   
1461   50          1      Travel_Rarely        410                   Sales   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
0                    1          2     Life Sciences              1   
2                    2          2             Other              1   
14                  24          3     Life Sciences              1   
21                   9          4     Life Sciences              1   
24                   6          1           Medical              1   
...                ...        ...               ...            ...   
1438                 9          3         Marketing              1   
1442                 1          4           Medical              1   
1444                 7          2  Technical Degree              1   
1452                 1          4     Life Sciences              1   
1461                28          3         Marketing              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
2                  4  ...                         2            80   
14                19  ...                         2            80   
21                27  ...                         2            80   
24                31  ...                         3            80   
...              ...  ...                       ...           ...   
1438            2023  ...                         1            80   
1442            2027  ...                         2            80   
1444            2032  ...                         4            80   
1452            2044  ...                         4            80   
1461            2055  ...                         2            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
2                    0                  7                      3   
14                   0                  6                      4   
21                   0                 10                      4   
24                   0                  8                      2   
...                ...                ...                    ...   
1438                 1                  1                      3   
1442                 3                  4                      3   
1444                 1                 14                      4   
1452                 2                 12                      3   
1461                 1                 20                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
2                  3               0                  0   
14                 3               4                  2   
21                 3               5                  3   
24                 3               4                  2   
...              ...             ...                ...   
1438               2               1                  0   
1442               4               2                  2   
1444               1              10               

In [7]:
pd.DataFrame(At_no)

Age  Attrition     BusinessTravel  DailyRate              Department  \
1      49          0  Travel_Frequently        279  Research & Development   
3      33          0  Travel_Frequently       1392  Research & Development   
4      27          0      Travel_Rarely        591  Research & Development   
5      32          0  Travel_Frequently       1005  Research & Development   
6      59          0      Travel_Rarely       1324  Research & Development   
...   ...        ...                ...        ...                     ...   
1465   36          0  Travel_Frequently        884  Research & Development   
1466   39          0      Travel_Rarely        613  Research & Development   
1467   27          0      Travel_Rarely        155  Research & Development   
1468   49          0  Travel_Frequently       1023                   Sales   
1469   34          0      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
1                    8          1  Life Sciences              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
5                    2          2  Life Sciences              1   
6                    3          3        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
1                  2  ...                         4            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
5                  8  ...                         3            80   
6                 10  ...                         1            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1                    1                 10                      3   
3                    0                  8                      3   
4                    1                  6                      3   
5                    0                  8                      2   
6                    3                 12                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
1                  3              10                  7   
3                  3               8                  7   
4                  3               2                  2   
5                  2               7                  7   
6                  2               1                  0   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2        

In [8]:
no_sample=At_no.sample(n=350)
no_sample

Age  Attrition     BusinessTravel  DailyRate              Department  \
635    35          0      Travel_Rarely        607  Research & Development   
139    30          0      Travel_Rarely       1240         Human Resources   
690    31          0      Travel_Rarely        616  Research & Development   
163    57          0      Travel_Rarely        334  Research & Development   
769    26          0  Travel_Frequently        921  Research & Development   
...   ...        ...                ...        ...                     ...   
948    30          0      Travel_Rarely        634  Research & Development   
1454   45          0      Travel_Rarely        374                   Sales   
1212   34          0      Travel_Rarely        678  Research & Development   
1004   26          0      Travel_Rarely        652  Research & Development   
164    27          0         Non-Travel       1450  Research & Development   

      DistanceFromHome  Education   EducationField  EmployeeCount  \
635                  9          3    Life Sciences              1   
139                  9          3  Human Resources              1   
690                 12          3          Medical              1   
163                 24          2    Life Sciences              1   
769                  1          1          Medical              1   
...                ...        ...              ...            ...   
948                 17          4          Medical              1   
1454                20          3    Life Sciences              1   
1212                19          3    Life Sciences              1   
1004                 7          3            Other              1   
164                  3          3          Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
635              880  ...                         2            80   
139              184  ...                         4            80   
690              961  ...                         3            80   
163              223  ...                         2            80   
769             1068  ...                         3            80   
...              ...  ...                       ...           ...   
948             1321  ...                         4            80   
1454            2046  ...                         3            80   
1212            1701  ...                         2            80   
1004            1417  ...                         4            80   
164              224  ...                         4            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
635                  1                 17                      2   
139                  0                 12                      2   
690                  2                 10                      2   
163                  1                 12                      2   
769                  2                  5                      5   
...                ...                ...                    ...   
948                  2                  9                      2   
1454                 0                  8                      3   
1212                 0                 10                      3   
1004                 0                  8                      2   
164                  1                  1                      2   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
635                3              17                 14   
139                1              11                  9   
690                1               9                  7   
163                1               5                  3   
769                3               5                  3   
...              ...             ...                ...   
948                3               9                  1   
1454               3               5                  3   
1212               3              10                  9   
1004

In [9]:
new_df = pd.concat([no_sample,At_yes],axis=0)
new_df

Age  Attrition     BusinessTravel  DailyRate              Department  \
635    35          0      Travel_Rarely        607  Research & Development   
139    30          0      Travel_Rarely       1240         Human Resources   
690    31          0      Travel_Rarely        616  Research & Development   
163    57          0      Travel_Rarely        334  Research & Development   
769    26          0  Travel_Frequently        921  Research & Development   
...   ...        ...                ...        ...                     ...   
1438   23          1  Travel_Frequently        638                   Sales   
1442   29          1      Travel_Rarely       1092  Research & Development   
1444   56          1      Travel_Rarely        310  Research & Development   
1452   50          1  Travel_Frequently        878                   Sales   
1461   50          1      Travel_Rarely        410                   Sales   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
635                  9          3     Life Sciences              1   
139                  9          3   Human Resources              1   
690                 12          3           Medical              1   
163                 24          2     Life Sciences              1   
769                  1          1           Medical              1   
...                ...        ...               ...            ...   
1438                 9          3         Marketing              1   
1442                 1          4           Medical              1   
1444                 7          2  Technical Degree              1   
1452                 1          4     Life Sciences              1   
1461                28          3         Marketing              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
635              880  ...                         2            80   
139              184  ...                         4            80   
690              961  ...                         3            80   
163              223  ...                         2            80   
769             1068  ...                         3            80   
...              ...  ...                       ...           ...   
1438            2023  ...                         1            80   
1442            2027  ...                         2            80   
1444            2032  ...                         4            80   
1452            2044  ...                         4            80   
1461            2055  ...                         2            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
635                  1                 17                      2   
139                  0                 12                      2   
690                  2                 10                      2   
163                  1                 12                      2   
769                  2                  5                      5   
...                ...                ...                    ...   
1438                 1                  1                      3   
1442                 3                  4                      3   
1444                 1                 14                      4   
1452                 2                 12                      3   
1461                 1                 20                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
635                3              17                 14   
139                1              11                  9   
690                1               9                  7   
163                1               5                  3   
769                3               5                  3   
...              ...             ...                ...   
1438               2               1                  0   
1442               4               2                  2   
1444               1              10               

In [10]:
new_df["Attrition"].value_counts()

Attrition
0    350
1    237
Name: count, dtype: int64

In [11]:
new_df

Age  Attrition     BusinessTravel  DailyRate              Department  \
635    35          0      Travel_Rarely        607  Research & Development   
139    30          0      Travel_Rarely       1240         Human Resources   
690    31          0      Travel_Rarely        616  Research & Development   
163    57          0      Travel_Rarely        334  Research & Development   
769    26          0  Travel_Frequently        921  Research & Development   
...   ...        ...                ...        ...                     ...   
1438   23          1  Travel_Frequently        638                   Sales   
1442   29          1      Travel_Rarely       1092  Research & Development   
1444   56          1      Travel_Rarely        310  Research & Development   
1452   50          1  Travel_Frequently        878                   Sales   
1461   50          1      Travel_Rarely        410                   Sales   

      DistanceFromHome  Education    EducationField  EmployeeCount  \
635                  9          3     Life Sciences              1   
139                  9          3   Human Resources              1   
690                 12          3           Medical              1   
163                 24          2     Life Sciences              1   
769                  1          1           Medical              1   
...                ...        ...               ...            ...   
1438                 9          3         Marketing              1   
1442                 1          4           Medical              1   
1444                 7          2  Technical Degree              1   
1452                 1          4     Life Sciences              1   
1461                28          3         Marketing              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
635              880  ...                         2            80   
139              184  ...                         4            80   
690              961  ...                         3            80   
163              223  ...                         2            80   
769             1068  ...                         3            80   
...              ...  ...                       ...           ...   
1438            2023  ...                         1            80   
1442            2027  ...                         2            80   
1444            2032  ...                         4            80   
1452            2044  ...                         4            80   
1461            2055  ...                         2            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
635                  1                 17                      2   
139                  0                 12                      2   
690                  2                 10                      2   
163                  1                 12                      2   
769                  2                  5                      5   
...                ...                ...                    ...   
1438                 1                  1                      3   
1442                 3                  4                      3   
1444                 1                 14                      4   
1452                 2                 12                      3   
1461                 1                 20                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
635                3              17                 14   
139                1              11                  9   
690                1               9                  7   
163                1               5                  3   
769                3               5                  3   
...              ...             ...                ...   
1438               2               1                  0   
1442               4               2                  2   
1444               1              10               

In [12]:
X = new_df.drop(["EmployeeCount","EmployeeNumber","Over18","StandardHours","Attrition"],axis=1)
y = new_df["Attrition"]
y

635     0
139     0
690     0
163     0
769     0
       ..
1438    1
1442    1
1444    1
1452    1
1461    1
Name: Attrition, Length: 587, dtype: int64

In [13]:
cat_col = X.select_dtypes(include=["object"]).columns.tolist()
num_col = X.select_dtypes(include=["int64","float64"]).columns.tolist()

In [14]:
preprocess = ColumnTransformer([
    ("onehot",OneHotEncoder(drop="first",sparse=False,handle_unknown="ignore"),cat_col),
    ("scaler",StandardScaler(),num_col)
])

In [15]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [16]:
pipeline=Pipeline([
    ("preprocessor",preprocess),
    ("classifier",RandomForestClassifier())
])

In [17]:
pipeline.fit(X_train,y_train)

C:\Users\Arhum\anaconda3\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('onehot',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse=False),
                                                  ['BusinessTravel',
                                                   'Department',
                                                   'EducationField', 'Gender',
                                                   'JobRole', 'MaritalStatus',
                                                   'OverTime']),
                                                 ('scaler', StandardScaler(),
                                                  ['Age', 'DailyRate',
                                                   'DistanceFromHome',
                                                   'Education',
                                                   'EnvironmentSatisfaction',
                                                   'HourlyRate'...
                                                   'JobSatisfaction',
                                                   'MonthlyIncome',
                                                   'MonthlyRate',
                                                   'NumCompaniesWorked',
                                                   'PercentSalaryHike',
                                                   'PerformanceRating',
                                                   'RelationshipSatisfaction',
                                                   'StockOptionLevel',
                                                   'TotalWorkingYears',
                                                   'TrainingTimesLastYear',
                                                   'WorkLifeBalance',
                                                   'YearsAtCompany',
                                                   'YearsInCurrentRole',
                                                   'YearsSinceLastPromotion',
                                                   'YearsWithCurrManager'])])),
                ('classifier', RandomForestClassifier())])

In [18]:
predictions=pipeline.predict(X_test)
predictions

array([1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 0, 0], dtype=int64)

In [19]:
from sklearn.metrics import classification_report, confusion_matrix

print(classification_report(y_test, predictions))
print(confusion_matrix(y_test, predictions))

              precision    recall  f1-score   support

           0       0.77      0.95      0.85        73
           1       0.86      0.53      0.66        45

    accuracy                           0.79       118
   macro avg       0.81      0.74      0.75       118
weighted avg       0.80      0.79      0.77       118

[[69  4]
 [21 24]]


This matrix shows the number of correct and incorrect predictions made by the classifier:

69: The number of times the model predicted class 0 when the true label was also class 0 (True Negative, TN).

4: The number of times the model predicted class 1 when the true label was class 0 (False Positive, FP).

21: The number of times the model predicted class 0 when the true label was class 1 (False Negative, FN).

24: The number of times the model predicted class 1 when the true label was class 1 (True Positive, TP).

Classification Report:
1. Precision:
Precision indicates how many of the instances the model predicted as positive (class 1) were actually positive.

Class 0: Precision is 0.77, meaning that 77% of the times the model predicted class 0, it was correct.

Class 1: Precision is 0.86, meaning that 86% of the times the model predicted class 1, it was correct.

2. Recall:
Recall (also known as sensitivity or true positive rate) tells us how many of the actual positives were correctly identified by the model.

Class 0: Recall is 0.95, meaning that 95% of all true class 0 labels were correctly identified as class 0 by the model.

Class 1: Recall is 0.53, meaning that only 53% of all true class 1 labels were correctly identified as class 1.

3. F1-Score:
The F1-score is the harmonic mean of precision and recall. It is useful when you need to balance both precision and recall.

Class 0: F1-score is 0.85.

Class 1: F1-score is 0.66.

4. Support:
Support represents the number of true instances for each class in the dataset.

Class 0: 73 samples

Class 1: 45 samples

Accuracy:
The overall accuracy is the percentage of correct predictions (both true positives and true negatives).

Accuracy = (TP + TN) / Total Samples = (69 + 24) / 118 = 0.79, or 79%.

Averages:
Macro Average: This is the average of precision, recall, and F1-score for each class, treating all classes equally.

Precision: 0.81

Recall: 0.74

F1-Score: 0.75

Weighted Average: This is the average of precision, recall, and F1-score for each class, weighted by the number of samples in each class. This helps in scenarios with imbalanced classes.

Precision: 0.80

Recall: 0.79

F1-Score: 0.77

Interpretation:
The model performs better on class 0 than on class 1. This is evident from the higher recall and precision for class 0.

Class 1 has a lower recall, meaning that the model misses many of the actual class 1 instances (higher false negatives).

The accuracy of the model is 79%, which is decent but suggests that the model might not be performing equally well for both classes.

The F1-score for class 1 (0.66) is lower, which could indicate that the model might not be doing a great job at detecting class 1, potentially due to the class imbalance (fewer samples for class 1).


